lib/mayaUsd/ufe/UsdConnectionHandler.cpp

```diff
@@ -88,30 +88,96 @@ bool UsdConnectionHandler::createConnection(
     const Ufe::Attribute::Ptr& srcAttr,
     const Ufe::Attribute::Ptr& dstAttr) const
 {
-    PXR_NS::UsdAttribute srcUsdAttr = usdAttrFromUfeAttr(srcAttr);
-    PXR_NS::UsdAttribute dstUsdAttr = usdAttrFromUfeAttr(dstAttr);
+    UsdAttribute* srcUsdAttr = usdAttrFromUfeAttr(srcAttr);
+    UsdAttribute* dstUsdAttr = usdAttrFromUfeAttr(dstAttr);
 
-    if (!isConnected(srcUsdAttr, dstUsdAttr)) {
-        return dstUsdAttr.AddConnection(srcUsdAttr.GetPath());
+    if (!srcUsdAttr || !dstUsdAttr
+        || isConnected(srcUsdAttr->usdAttribute(), dstUsdAttr->usdAttribute())) {
+        return false;
     }
 
-    // Return a failure because there is already a connection.
+    UsdShadeConnectableAPI srcApi(srcUsdAttr->usdPrim());
+    TfToken                srcBaseName;
+    UsdShadeAttributeType  srcAttrType;
+    std::tie(srcBaseName, srcAttrType)
+        = UsdShadeUtils::GetBaseNameAndType(TfToken(srcAttr->name()));
+
+    UsdShadeConnectableAPI dstApi(dstUsdAttr->usdPrim());
+    TfToken                dstBaseName;
+    UsdShadeAttributeType  dstAttrType;
+    std::tie(dstBaseName, dstAttrType)
+        = UsdShadeUtils::GetBaseNameAndType(TfToken(dstAttr->name()));
+
+    if (srcAttrType == UsdShadeAttributeType::Input) {
+        UsdShadeInput srcInput = srcApi.CreateInput(srcBaseName, srcUsdAttr->usdAttributeType());
+        if (dstAttrType == UsdShadeAttributeType::Input) {
+            UsdShadeInput dstInput
+                = dstApi.CreateInput(dstBaseName, dstUsdAttr->usdAttributeType());
+            return UsdShadeConnectableAPI::ConnectToSource(dstInput, srcInput);
+        } else {
+            UsdShadeOutput dstOutput
+                = dstApi.CreateOutput(dstBaseName, dstUsdAttr->usdAttributeType());
+            return UsdShadeConnectableAPI::ConnectToSource(dstOutput, srcInput);
+        }
+    } else {
+        UsdShadeOutput srcOutput = srcApi.CreateOutput(srcBaseName, srcUsdAttr->usdAttributeType());
+        if (dstAttrType == UsdShadeAttributeType::Input) {
+            UsdShadeInput dstInput
+                = dstApi.CreateInput(dstBaseName, dstUsdAttr->usdAttributeType());
+            return UsdShadeConnectableAPI::ConnectToSource(dstInput, srcOutput);
+        } else {
+            UsdShadeOutput   dstOutput;
+            UsdShadeMaterial dstMaterial(dstUsdAttr->usdPrim());
+            // Special case when connecting to Material outputs:
+            if (dstMaterial
+                && (dstBaseName == UsdShadeTokens->surface || dstBaseName == UsdShadeTokens->volume
+                    || dstBaseName == UsdShadeTokens->displacement)) {
```

Comment:
Can that list of `materials` change (e.g. new material type with MatX, Arnold, etc.)?

Reply:
Discussed, and the answer is no.